### PR TITLE
Bring test coverage above 95%

### LINKED ISSUE #438: 100 coverage
We need 100 coverage.

- [x] `maxHeadersSize` option is missing tests.
- [x] `Agent.closed`, `Agent.destroyed` is missing tests.
- [ ] redirect to `onUpgrade`

```
npm run coverage
```

### PR DESCRIPTION
**Pull request by ShogunPanda**
This PR provides mainly two things:

1. Make sure it is possible to connect to IPv6 IPs. @mcollina @ronag Please double check this. I randomly found this while bringing coverage up.
2. Bring coverage above 95% (now it is around 98%).

### CODE REVIEW

File: index.js
Comment (mcollina) [inline index.js:35]: why ignoring? Could this even happen?

**Review by mcollina — COMMENTED** [review]
Can you please add the -100 check of tap so we do not regress?

File: index.js
Comment (ronag) [reply index.js:35]: This can't happen. Just remove the `||`.

File: lib/agent/redirect.js
Comment (ronag) [inline lib/agent/redirect.js:19]: All if these can happen. Please put them back.

File: lib/agent/redirect.js
Comment (ronag) [inline lib/agent/redirect.js:37]: Just remove the `||`.

File: lib/agent/redirect.js
Comment (ronag) [inline lib/agent/redirect.js:95]: This can happen. Please put back.

File: lib/client-pool.js
Comment (ronag) [inline lib/client-pool.js:107]: This should be possible to happen. Please remove ignore.

File: lib/core/client.js
Comment (ronag) [inline lib/core/client.js:68]: This can happen. Please remove ignore.

**Review by ronag — CHANGES_REQUESTED** [review]
Excellent work. Please fix the comments and remove all the istanbul ignores I think some of them we can make tests for. I'd like to have a shot at them as well later.

File: lib/agent/redirect.js
Comment (ShogunPanda) [reply lib/agent/redirect.js:95]: I'm not sure I understand.
In the comment I'm not saying it cannot happen, I'm saying that we don't handle them, like we do with the body.

I just forgot to remove the `this.agent.dispatch` line below. Do you agree or do you explicitily want to handle trailers?

File: index.js
Comment (ShogunPanda) [reply index.js]: Done!

File: lib/agent/redirect.js
Comment (ShogunPanda) [reply lib/agent/redirect.js]: Done. I was not able to find in the code how to reach those hooks so I leave it up to you.

File: lib/agent/redirect.js
Comment (ShogunPanda) [reply lib/agent/redirect.js]: Done!

File: lib/client-pool.js
Comment (ShogunPanda) [reply lib/client-pool.js:107]: Done. I leave it to you to add a test for it. I've tried with no luck.

File: lib/core/client.js
Comment (ShogunPanda) [reply lib/core/client.js:68]: How comes? Since we are on Node LTS (so > 10 at the moment), `client[kUrl]` is a instance of `URL` (the WHATWG one), so if `host` is not defined, neither is `hostname`.

Am I wrong?

**Comment by ShogunPanda** [comment]
@mcollina I'm not sure I can at the moment. Since Robert has to provide missing tests for the cases I unignored, if I do that now the CI will fail. Unless we can agree for a lower threshold, like 95%.

@ronag I've made most of the fixes you asked for, except for two I left comment for. Do you want me to remove ALL remaining istanbul ignores? Just double checking.

File: lib/core/client.js
Comment (ronag) [reply lib/core/client.js:68]: you are right, then please remove the `||` part.

File: lib/agent/redirect.js
Comment (ronag) [reply lib/agent/redirect.js:95]: I would add a comment so that it's explicit that we ignore them on purpose

File: lib/agent/index.js
Comment (ronag) [inline lib/agent/index.js:97]: This can happen? It's just the data event that we don't bother testing for.

File: lib/agent/redirect.js
Comment (ShogunPanda) [reply lib/agent/redirect.js:95]: Ok, and then I'll redirect to the full explanation above. Understood. Updated coming soon.

File: lib/client-pool.js
Comment (ronag) [reply lib/client-pool.js:107]: Please remove the ignore

**Review by ronag — CHANGES_REQUESTED** [review]
Please remove all the ignores. I'll have a go at it later.

File: lib/core/client.js
Comment (ronag) [inline lib/core/client.js:878]: remove ternary and just make an `assert(idx !== -1)`

File: lib/core/client.js
Comment (ronag) [inline lib/core/client.js:883]: is the else possible here? what happens then?

File: lib/core/util.js
Comment (ronag) [inline lib/core/util.js:170]: Could we just do `global.queueMicrotask = null` in a test?

File: test/agent.js
Comment (ronag) [inline test/agent.js:491]: also try negative integer

File: lib/agent/index.js
Comment (ShogunPanda) [reply lib/agent/index.js:97]: If you look at line 90, there is an entire check about body  empty stream. So, the if on 98 is only satisfied by 
I was not able to simulate a empty stream which emits so I skipped the if check as well.

Comment (ronag) [reply lib/agent/index.js:97]: I guess it doesn't matter so much.

File: lib/core/client.js
Comment (ShogunPanda) [reply lib/core/client.js:878]: It should not because URL parsing would fail.
Ok, will do that!

File: lib/core/client.js
Comment (ShogunPanda) [reply lib/core/client.js:883]: Same thing. I think I will add an assert here as well. Else should not be possible.

File: lib/agent/redirect.js
Comment (ronag) [inline lib/agent/redirect.js:92]:
```suggestion
        no body is generated (and thus can be ignored). Even though generating a body is not prohibited.
```

File: lib/client-pool.js
Comment (ShogunPanda) [reply lib/client-pool.js:107]: Update coming soon!

File: lib/core/client.js
Comment (ShogunPanda) [reply lib/core/client.js:68]: Update coming soon!

File: lib/core/util.js
Comment (ShogunPanda) [reply lib/core/util.js:170]: Mh...we could but I definitely don't like playing with global.
Also, since we only support Node and we are on >10, do we really need this?

File: test/agent.js
Comment (ShogunPanda) [reply test/agent.js:491]: Good point. On my way!

File: lib/core/util.js
Comment (mcollina) [reply lib/core/util.js:170]: I would remove this completely. `queueMicrotask` should be available in 12+, and we should stop support v10 in the next major anyway.

File: lib/core/client.js
Comment (ronag) [reply lib/core/client.js:878]: I'm fine with doing istanbul ignore or assertions.

File: lib/core/util.js
Comment (ronag) [reply lib/core/util.js:170]: I'm good with that

Comment (ronag) [reply lib/core/util.js:170]: Removed in separate PR. Ignore it here.

File: lib/agent/redirect.js
Comment (ShogunPanda) [reply lib/agent/redirect.js:95]: Done!

File: test/agent.js
Comment (ShogunPanda) [reply test/agent.js:491]: Done!

File: lib/client-pool.js
Comment (ShogunPanda) [reply lib/client-pool.js]: Done!

File: lib/core/client.js
Comment (ShogunPanda) [reply lib/core/client.js]: Done!

File: lib/core/client.js
Comment (ShogunPanda) [reply lib/core/client.js]: Done!

File: lib/core/client.js
Comment (ShogunPanda) [reply lib/core/client.js]: Done!

File: lib/agent/redirect.js
Comment (ShogunPanda) [reply lib/agent/redirect.js]: Thanks for noticing. Fixed.

**Comment by ShogunPanda** [comment]
@mcollina `-100` was not possible. For now I've added `--lines 95 --branches 95 --functions 95 --statements 95` so we have a good starting point

@ronag I've removed all the unwanted ignores and update this PR description accordingly. I'll leave it to you to write test for the missing cases (or reinstate `ignores` ;)). 

Once this is merged I think also #603 can be merged, isn't it?

**Pull request merged**
